### DISCUSSION TITLE
Gradient calculation of fixed point does not converge

### DISCUSSION THREAD

**Discussion by JanLuca** [jax-ml/jax#12271] [Q&A]
Dear jax community,

I encountered a problem in my custom VJP gradient calculations of a fixed point iteration. The implementation follows the description in [1] and I already used the custom rule in other fixed point problems without any problem.

For my current problem I have a fixed point iteration whose body part of the while loop are along the line (omitting reshapes and similar things):

```python
def body_func(carry):
    A, B, old_result_1, old_result_2 = carry

    right_side_1 = jnp.tensordot(A, old_result_2.conj(), ...)
    left_side_1 = jnp.tensordot(old_result_2, right_side_1, ...)
    new_result_1 = jnp.linalg.solve(left_side_1, right_side_1)

    right_side_2 = jnp.tensordot(B, new_result_1.conj(), ...)
    left_side_2 = jnp.tensordot(new_result_1, right_side_2, ...)
    new_result_2 = jnp.linalg.solve(left_side_2, right_side_2)

    return A, B, new_result_1, new_result_2
```

Since I need the gradient of `A` and `B` I implemented a custom VJP for the fixed point iteration. But here I encounter now the problem that the fixed point iteration in the forward pass converges without any problem and I found elementwise converged arrays `result_1` and `result_2` but in the reverse pass the fixed point iteration needed to calculate the gradient does not converge. The difference between two steps hangs around a value 10⁻⁴ and do not reach my threshold of 10⁻⁸.

I want to ask now if it is known that this instability of the gradient calculation is common and if there is a general good solution to it. Or is this maybe a problem of the gradient of the linear solver (`jnp.linalg.solve`) that the VJP of it can be numerically instable for specific input (matrices with small singular values, ...) and thus leads to the instability of the VJP of the fixed point iteration?

I am thankful for any help and can provide of course more details if needed!

[1] https://jax.readthedocs.io/en/latest/notebooks/Custom_derivative_rules_for_Python_code.html#implicit-function-differentiation-of-iterative-implementations
